### PR TITLE
Additional ExpenseSheetCalculators

### PR DESCRIPTION
**Pull request by skius**

### CODE REVIEW

File: api/app/models/expense_sheet.rb [hunk 1]
```diff
@@ -41,6 +41,9 @@ class ExpenseSheet < ApplicationRecord
 
   scope :before_date, (->(date) { where(arel_table[:ending].lt(date)) })
 
+  # ExpenseSheets which can be used in calculations
+  scope :calculable, -> { where.not(state: :open) }
```
Comment (lukasbischof) [inline api/app/models/expense_sheet.rb]: Maybe `relevant_for_calculation`

File: api/spec/models/service_spec.rb [hunk 2]
```diff
@@ -9,6 +9,38 @@
...
+    let(:used_days_calculator) { instance_double ExpenseSheetCalculators::UsedDaysCalculator }
+    let(:remaining_days_calculator) { instance_double ExpenseSheetCalculators::RemainingDaysCalculator }
+
+    before do
+      allow(ExpenseSheetCalculators::UsedDaysCalculator).to receive(:new).and_return used_days_calculator
+      allow(ExpenseSheetCalculators::RemainingDaysCalculator).to receive(:new).and_return remaining_days_calculator
+    end
+
+    it 'creates a used_days_calculator' do
+      service.send(:used_days_calculator)
```
Comment (lukasbischof) [inline api/spec/models/service_spec.rb]: Avoid using send

File: api/spec/services/expense_sheet_calculators/expenses_calculator_spec.rb [hunk 3]
```diff
@@ -2,8 +2,8 @@
 
 require 'rails_helper'
 
-RSpec.describe ExpenseSheetCalculatorService, type: :service do
-  let(:calculator) { ExpenseSheetCalculatorService.new(expense_sheet) }
+RSpec.describe ExpenseSheetCalculators::ExpensesCalculator, type: :service do
+  let(:calculator) { ExpenseSheetCalculators::ExpensesCalculator.new(expense_sheet) }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/expenses_calculator_spec.rb]:
```suggestion
  let(:calculator) { described_class.new(expense_sheet) }
```

File: api/spec/services/expense_sheet_calculators/remaining_days_calculator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ExpenseSheetCalculators::RemainingDaysCalculator, type: :service do
+  let(:calculator) { ExpenseSheetCalculators::RemainingDaysCalculator.new(service) }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/remaining_days_calculator_spec.rb]:
```suggestion
  let(:calculator) { described_class.new(service) }
```

File: api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,176 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ExpenseSheetCalculators::SuggestionsCalculator, type: :service do
+  let(:calculator) { ExpenseSheetCalculators::SuggestionsCalculator.new(expense_sheet) }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb]:
```suggestion
  let(:calculator) { described_class.new(expense_sheet) }
```

File: api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  let(:beginning) { Date.parse('2018-01-01') }
+  let(:ending) { Date.parse('2018-01-26') }
+  let!(:service) { create :service, beginning: beginning, ending: ending, user: user }
+  let(:expense_sheet) { create :expense_sheet, beginning: beginning, ending: ending, user: user }
+
+  let(:expected_work_days) { 20 }
+  let(:expected_workfree_days) { 6 }
+
+  describe '#suggested_work_days' do
+    let(:day_calculator) { instance_double(DayCalculator) }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb]:
```suggestion
    let(:day_calculator) { instance_double(DayCalculator, calculate_work_days: expected_work_days) }
```

File: api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    before do
+      allow(DayCalculator).to receive(:new).and_return day_calculator
+      allow(day_calculator).to receive(:calculate_work_days).and_return expected_work_days
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb]: You may be able to remove this line

File: api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+  describe '#suggested_workfree_days' do
+    let(:day_calculator) { instance_double(DayCalculator) }
+
+    before do
+      allow(DayCalculator).to receive(:new).and_return day_calculator
+      allow(day_calculator).to receive(:calculate_workfree_days).and_return expected_workfree_days
+    end
+
+    it 'delegates the correct method' do
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb]:
```suggestion
    it 'delegates the correct method', :aggregate_failures do
```

File: api/spec/services/expense_sheet_calculators/used_days_calculator_spec.rb [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  let(:beginning) { service_range.begin }
+  let(:ending) { service_range.end }
+  let(:service) { create :service, beginning: beginning, ending: ending }
+  let(:created_expense_sheets) { ExpenseSheetGenerator.new(service).create_expense_sheets }
+
+  describe '#used_sick_days' do
+    subject { calculator.used_sick_days }
+
+    context 'with no relevant_for_calculations expense_sheets' do
+      before { created_expense_sheets.each { |expense_sheet| expense_sheet.update sick_days: 10 } }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/used_days_calculator_spec.rb]: Should work as well:
```suggestion
      before { created_expense_sheets.update sick_days: 10 }
```

File: api/spec/services/expense_sheet_calculators/used_days_calculator_spec.rb [hunk 10]
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ExpenseSheetCalculators::UsedDaysCalculator, type: :service do
+  let(:calculator) { ExpenseSheetCalculators::UsedDaysCalculator.new(service) }
```
Comment (lukasbischof) [inline api/spec/services/expense_sheet_calculators/used_days_calculator_spec.rb]:
```suggestion
  let(:calculator) { described_class.new(service) }
```

File: api/spec/models/service_spec.rb (hunk 2 above)
Comment (skius) [reply api/spec/models/service_spec.rb]: Resolved by calling the delegated methods

File: api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb (hunk 7 above)
Comment (skius) [reply api/spec/services/expense_sheet_calculators/suggestions_calculator_spec.rb]: Yes!

**Pull request merged**